presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlockEncoding.java

```diff
@@ -45,7 +46,8 @@ public void writeBlock(BlockEncodingSerde blockEncodingSerde, SliceOutput sliceO
         blockEncodingSerde.writeBlock(sliceOutput, dictionary);
 
         // ids
-        sliceOutput.writeBytes(dictionaryBlock.getIds());
+        ImmutableIntArray ids = dictionaryBlock.getIds();
+        writeIds(sliceOutput, ids.getArray(), ids.getOffset(), ids.getLength());
```

Comment:
Can SliceOutput handle more elements when Slice? If not, this change won't help here.

Reply:
@mbasmanova  SliceOutput does not support >2GB size, but the change works here, because 
1) OptimizedPartitionedOutputOperator would cut the input block and append to buffer in batches. For this particular issue, the input block to OptimizedPartitionedOutputOperator after block flattening is a ColumnarArray like this:
```
    offsets: 1000 elements: [0, 8, 10, 1294, 2578, 3862,...500000000]
    elementsBlock: DictionaryBlock of many many repeating ids(could be > 500M):
                                     ids: 500M [48,49,50,51,52,53,54,55, 165,166,  288,289,...1571,  288,289,...1571, .....]
```
Each of the array in ColumnarArray only contain thousands not millions of elements. OptimizedPartitionedOutputOperator will only append the data that can fit into buffer.

2) DictionaryBlockEncoding is used by old PartitionedOutputOperator, which does not do block flattening, and thus doesn't have this problem.

Reply:
@yingsu00 Thanks for explaining. Would it make sense to keep existing getIds method as is and introduce `int[] getRawIds()` method to be used in BlockFlattener? This removes the need for a new class. (BTW, the name ImmutableIntArray is misleading because the array is actually mutable: x.getArray()[5] = 10 works.)